source/Timer.cpp

```diff
@@ -0,0 +1,322 @@
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Timer.h"
+
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "GameData.h"
+#include "Logger.h"
+#include "Planet.h"
+#include "PlayerInfo.h"
+#include "Random.h"
+#include "Ship.h"
+#include "StellarObject.h"
+#include "UI.h"
+
+using namespace std;
+
+Timer::Timer(const DataNode &node, const Mission *mission)
+{
+	Load(node, mission);
+}
+
+
+
+void Timer::Load(const DataNode &node, const Mission *mission)
+{
+	if(node.Size() > 1)
+		name = node.Token(1);
+
+	this->mission = mission;
+
+	for(const DataNode &child : node)
+	{
+		if(child.Token(0) == "time" && child.Size() > 1)
+		{
+			base = static_cast<int64_t>(child.Value(1));
+			if(child.Size() > 2)
+				rand = static_cast<uint32_t>(child.Value(2));
+		}
+		else if(child.Token(0) == "idle")
+			requireIdle = true;
+		else if(child.Token(0) == "optional")
+			optional = true;
+		else if(child.Token(0) == "system" && child.Size() > 1)
+			system = GameData::Systems().Get(child.Token(1));
+		else if(child.Token(0) == "system")
+			systems.Load(child);
+		else if(child.Token(0) == "proximity")
+		{
+			// If the proximity node specifies a single planet, we take that;
+			// otherwise, we look to its child for a LocationFilter
+			if(child.Size() > 1)
+				proximityCenter = GameData::Planets().Find(child.Token(1));
+			else if(child.HasChildren())
+			{
+				const DataNode &firstGrand = (*child.begin());
+				proximityCenters.Load(firstGrand);
+			}
+		}
+		else if(child.Token(0) == "proximity settings")
+		{
+			if(child.Size() > 1)
+				proximity = child.Value(1);
+			if(child.Size() > 2)
+			{
+				if(child.Token(2) == "close")
+					closeTo = true;
+				else if(child.Token(2) == "far")
+					closeTo = false;
+			}
+		}
+		else if(child.Token(0) == "uncloaked")
+			requireUncloaked = true;
+		else if(child.Token(0) == "reset")
+		{
+			if(child.Size() > 1 && child.Token(1) == "leave system")
+				resetCondition = ResetCondition::LEAVE_SYSTEM;
+			else if(child.Size() > 1 && child.Token(1) == "leave zone")
+				resetCondition = ResetCondition::LEAVE_ZONE;
+			else if(child.Size() > 1 && child.Token(1) == "none")
+				resetCondition = ResetCondition::NONE;
+			else if(child.Size() > 1 && child.Token(1) == "pause")
+				resetCondition = ResetCondition::PAUSE;
+		}
+		else if(child.Token(0) == "repeat reset")
+			repeatReset = true;
+		else if(child.Token(0) == "reset fired")
+			resetFired = true;
+		// We keep "on timeup" as separate tokens so that it's compatible with MissionAction syntax
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "timeup")
+			action.Load(child);
+		else if(child.Token(0) == "on" && child.Size() > 1 && child.Token(1) == "reset")
+			resetAction.Load(child);
+
```

Comment:
```suggestion
			resetAction.Load(child);
		else
			child.PrintTrace("Skipping unrecognized attribute:");
```

Reply:
(Adding this locally because I changed the previous line)